### PR TITLE
bloomon-translation review

### PR DESCRIPTION
**Pull request by dinokong0128**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,14 +14,14 @@
   "author": "Dino Kong",
   "license": "ISC",
   "dependencies": {
+    "body-parser": "^1.17.1",
     "bookshelf": "^0.10.3",
-    "cps": "^1.0.2",
-    "express": "^4.15.2",
     "knex": "^0.12.7",
     "morgan": "^1.8.1",
     "node-cache": "^4.1.1",
```
Comment (williamriancho) [inline package.json]: Is this still needed?

File: src/config.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  user: process.env.DB_USER || 'root',
+  password: process.env.DB_PASSWORD || 'root',
+  database: process.env.DB_NAME || 'bloomon_translation_testing',
+  dbType: 'mysql',
+  port: process.env.PORT || 3030,
+
+  
+
+  cacheExpireTime: 600, // in seconds.
+  cacheCheckInterval: 300, // in seconds.
```
Comment (williamriancho) [inline src/config.js]: Is this still needed?

File: src/server.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        .catch((err) => {
+          return next(new restify.errors.InternalServerError('db error.'));
+        });
+    }
+  });
+});
+
+app.post('/webhooks',
+  (req, res, next) => {
+    if (webhookValidate(req.header('X-PhraseApp-Signature', ''), JSON.stringify(req.body))) {
```
Comment (williamriancho) [inline src/server.js]: Create a dedicated middleware (`/middlewares/phrase-app-auth.js` for instance)

File: src/server.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+const cache = require('./services/node_cache');
+const Translation = require('./models/translation');
+const reloadDB = require('./services/reload_db');
+const webhookValidate = require('./services/webhook_validate');
+
+const app = restify.createServer({name:''});
+
+app.use(morgan(env === 'PRODUCTION' ? 'combined' : 'dev'));
+app.use(restify.queryParser());
+app.use(bodyParser.json());
```
Comment (williamriancho) [inline src/server.js]: Use it only if needed

File: src/config/server.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  env: process.env.NODE_ENV || 'DEV',
+  port: process.env.PORT || 3030,
+
+  cacheExpireTime: 600, // in seconds.
+  cacheCheckInterval: 300, // in seconds.
```
Comment (williamriancho) [inline src/config/server.js]: Is this needed?

File: src/server.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+});
+
+app.on('InternalServer', (req, res, err, cb) => {
+  console.log(err);
+  cb();
+});
+
+app.listen(port, () => {
+  console.log(`Listen on port: ${port}`);
+});
```
Comment (williamriancho) [inline src/server.js]:
```js
const server = app.listen(port, () => {
	const addr = server.address();
        console.log(`Listening on ${addr.address}:${addr.port}`);
});
```

File: src/services/webhook_validate.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+const {env} = require('../config/server');
```
Comment (williamriancho) [inline src/services/webhook_validate.js]: Move this to the phrase app auth middleware

File: src/server.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+      return next();
+    } else {
+      return next(new restify.errors.BadRequestError('invalid request.'));
+    }
+  },
+  (req, res, next) => {
+    const data = req.body;
+    switch(data.event) {
+      case 'translations:create':
+      case 'translations:update':
```
Comment (williamriancho) [inline src/server.js]: What about deletion?

File: src/server.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+              )
+            })
+            .then(next())
+            .catch((err) => {
+              return next(new restify.errors.InternalServerError('db error.'));
+            });
+        }
+        break;
+      default:
+        return next();
```
Comment (williamriancho) [inline src/server.js]: You probably want to pass an error to next here.

**Comment by williamriancho** [comment]
Can you use dashes instead of underscores for the file names.

File: src/server.js (hunk 8 above)
Comment (dinokong0128) [inline src/server.js]: I don't see an event for that.
https://phraseapp.com/docs/guides/working-with-phraseapp/webhooks/
locales:create - A new Language Version has been created on a project
locales:update - A Language Version has been changed, renamed or reconfigured
locales:delete - A Language Version has been deleted
uploads:create - A Locale File has been uploaded
uploads:processing - A Locale File was processed
keys:create - A Key has been created
keys:update - A Key has been renamed or changed
keys:delete - A Key has been deleted
keys:batch_delete - Multiple Keys were deleted
translations:create - A Translation of a Key in a certain Language Version has been added
translations:update - A Translation of a Key in a certain Language Version has been edited
translations:deliver - A Translation of a Key in a certain Language Version has been delivered by a language service provider
comments:create - A Comment on a Translation Key has been added

**Comment by williamriancho** [comment]
Can you group the configuration in a yaml file:

```js
// config.js
const yaml =  require('js-yaml');
const fs = require('fs');

const configFile = process.env.CONFIG || 'config.yml';
const config = yaml.safeLoad(fs.readFileSync(configFile));
module.exports = config;
```

**Comment by williamriancho** [comment]
Questions:
- Do we still need to invalidate the cache if we get the update from phraseapp?

**Comment by dinokong0128** [comment]
[[translation integration] microservice for translation](https://trello.com/c/kggvgrfH/514-translation-integration-microservice-for-translation)

File: src/server.js (hunk 4 above)
Comment (dinokong0128) [inline src/server.js]: it is needed otherwise req.body will be undefined.
cf: http://expressjs.com/en/4x/api.html#req.body

**Comment by dinokong0128** [comment]
@williamriancho hi William I just finished modifying according to your comment (except some i don't agree with which i had replied to you under the comments) and done refactoring the entire micro service to use Redis. please let me know how you think. thank you!

**Comment by williamriancho** [comment]
Move the `config.yml` at the root of the repo

**Comment by williamriancho** [comment]
Move the `test` directory at the root of the repo

**Comment by williamriancho** [comment]
`redis.createClient` calls should be only in the redis service

File: src/services/retrieve-data.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+const fetchData = require('./fetch-data');
+
+const redisClient = redis.createClient(port, host);
+
+module.exports = (locale, group, forceFetch=false) => {
+  const cacheKey = getCacheKey(locale, group);
+  return redisClient.getAsync(cacheKey)
+    .then((cachedData) => {
+      if (cachedData && !forceFetch) {
+        return JSON.parse(cachedData);
```
Comment (williamriancho) [inline src/services/retrieve-data.js]: `return` or `else` not both

**Pull request closed without merging**